### PR TITLE
Update CQUI hotkey handling

### LINKED ISSUE #22: idea - support hotkeys configuration for Air units - range attack and rebase
one of the things that makes the late game painful for hands is extra mouse clicks for Air unit range attacks and rebase operations. there is a hotkey for a range attack, but it conflicts with "camera left".
let's see if we can allow these things to be configurable in UI:
* range attack for air units
* rebase for air units

### LINKED ISSUE #40: Investigate proper use for Air Attack and WMD with Keyboard shortcuts
The soothsayer fix disabled the hot keys for Air Attack and WMD shortcuts because those they require additional parameters that were not being passed along.  Preferred to defer the fixes for these as it's not obvious how useful either of those commands are when they require pointing at something to attack

### LINKED ISSUE #96: Removing some CQUI hotkeys
I would like to remove CQUI handling of some keyboard shortcuts.
These seem to be all the keys that CQUI support and also can be rebound to whatever the user wants in the base game.

Function|Base game key default|CQUI
-|-|-
Next City|]|E (When city selected)
Previous City|[|Q (When city selected)
Next Ready Unit|.|E (When city not selected)
Prev Ready Unit|,|Q (When city not selected)
Pan Map|Arrow Keys|WSAD
Auto Explore|E|Alt + E
Capital City|\ |Home
Air Strike/Ranged Attack|R|S or Alt+S depending on CQUI option

I think CQUI added these since originally the game didn't have hotkeys for some of these functions. For some that had assigned hotkeys the game didn't allow remapping to other keys (I believe panning was hardcoded to the arrow keys).
Since now the base game does have the functionality to remap those functions I am proposing that we remove them from CQUI and let the base game handle them. Any thoughts, objections, concerns? (@alexeyOnGitHub , @the-m4a , @Infixo , and anyone else who runs across this)

### PR DESCRIPTION
**Pull request by alimulhuda**
Fixes #96 
Fixes #40 
Fixes #22 (The air attack is handled by base game and should be `R` by default)
Updated the Bindings screen to display the key bindings handled by CQUI in Classic mode (Bindings are currently always displayed but are only handled in Classic mode)
Allows remapping using the `cqui_settings_local.sql` file

image

Tested on Gathering Storm (New game and loading an existing game)

### CODE REVIEW

File: Assets/UI/worldinput_CQUI.lua [hunk 1]
```diff
@@ -14,22 +14,37 @@ BASE_CQUI_OnUserRequestClose       = OnUserRequestClose;
...
 LuaEvents.CQUI_SettingsUpdate.Add( CQUI_OnSettingsUpdate );
 LuaEvents.CQUI_SettingsInitialized.Add( CQUI_OnSettingsUpdate );
 
+-- ===========================================================================
+function CQUI_GetActionFromKey( uiKey:number )
+    local action :table = {};
+    for i, keymap in ipairs(CQUI_keyMaps) do
+        if (keymap["Alt"] == CQUI_isAltDown) and (keymap["Shift"] == CQUI_isShiftDown) and (Keys[keymap["Key"]] == uiKey) then
+            action[keymap["Action"]] = true;
+        end
```
[the remark this answers is not in the document]
Comment (alimulhuda) [reply Assets/UI/worldinput_CQUI.lua:43]: The absence of break is intentional. There are some bindings which are situational (F is for fishing boat and fort, N for mine and nuke, P for pasture and plantation). I did not want to force the keys for those functions to always be the same, so I made it such that the function will return a table with multiple values for all the possible actions the key should trigger. I don't think there should be a significant performance penalty for a simple loop search in this use case, so I went with the ability to have multiple functions mapable to one key instead.

File: Assets/UI/worldinput_CQUI.lua [hunk 2]
```diff
@@ -148,201 +133,116 @@ function DefaultKeyUpHandler( uiKey:number )
...
+
+    local bCanStartRebase = selectedUnit and UnitManager.CanStartOperation(selectedUnit, UnitOperationTypes.REBASE, nil, true);
+    if (action["REBASE"] and bCanStartRebase) then
+        -- Valid plot search code based on OnInterfaceModeChange_ReBase in WorldInput.lua
+        local tResults = UnitManager.GetOperationTargets(selectedUnit, UnitOperationTypes.REBASE );
+        local tValidPlots = tResults[UnitOperationResults.PLOTS];
+        if (tValidPlots ~= nil and table.count(tValidPlots) > 0) then
+            UI.SetInterfaceMode(InterfaceModeTypes.REBASE);
+        end
+        cquiHandledKey = true;
```
[the remark this answers is not in the document]
Comment (alimulhuda) [reply Assets/UI/worldinput_CQUI.lua:244]: If you find it more readable I'll try to follow that style. I have a tendency to put line breaks depending on situation, like I'll skip a line break if it means keeping together things that I think should be part of the same logical block. I might sometimes forget, so if you notice it just let me know and I'll modify to match that style.

**Comment by alimulhuda** [comment]
> Should we stick with the "Standard" and "Classic"?  I always thought that was kinda confusing anyway.  If "Standard" is "Civ6 Default", we should call it Civ6 Default.. "Classic" can become "CQUI Enhanced Keys" or something.

I think it's based on classic meaning Civ5 bindings. We can change the text if you think it makes it clearer. I'll go with what you suggested. 

> Can the Keys be hidden when the "Standard" (or "Not CQUI") setting is selected?

Should be possible. Hadn't done it because it got late and I got lazy. I'll see if i can get it to work nicely.

> we can put some "validate the loaded value is legal" code in the CQUICommon.lua file

I'll give it a try and update.

> I did verify that changing a value in the SQL file (I changed farm to "J") and starting a new game will correctly see that key as J.

I made it so that the key bindings should always be read from the database object (GameInfo) instead of the save game (GameConfiguration). So it should also work if you modify the key and load up an old game.

File: Assets/UI/worldinput_CQUI.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (the-m4a) [reply Assets/UI/worldinput_CQUI.lua:43]: ooh - my bad for misreading that design.  nice!

**Pull request merged**